### DISCUSSION TITLE
How to cache docker images in infrastructure-deployer pipeline

### DISCUSSION THREAD

**Discussion by zackleman** [gruntwork-io/knowledge-base#627] [How do I]
I wanted to speed up the build of docker containers so I switched the task launch_type to EC2 for infrastructure-deployer. A new EC2 cluster was spun up. Everything is working, however, the image layers are not being cached and reused so there is no speed-up in the build. This may be because kaniko is still being used. Is there a way to have the docker build not use kankiko so caching is enabled or a way to enable kaniko caching to take advantage of the new persistent storage with the EC2 launch type?

---

<ins datetime="2023-01-09T17:30:25Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109777">Tracked in ticket #109777</a></p>
</ins>

**Comment by gitsstewart** [comment]
@zleman1593 have you already reviewed https://discuss.circleci.com/t/how-to-build-and-push-docker-images-with-kaniko/44923 and tried implementing kaniko's caching feature https://github.com/GoogleContainerTools/kaniko#caching ?

**Reply by zackleman** [reply]
Yes I see caching in Kaniko is an option. I'm hoping someone from gruntwork can respond on how to best enable caching either with kaniko in the gruntwork ci/cd pipeline (since the gruntwork docs say enabling EC2 allows cacheing) or bypass it and just use docker build on the EC2 instance.

**Reply by denis256** [reply]
Hi, AFAIK Kaniko caching is in queue to be implemented https://github.com/gruntwork-io/terraform-aws-ci/issues/488

**Comment by zackleman** [comment]
I ended up getting some caching enabled by upgrading the kaniko docker image to use F`ROM gcr.io/kaniko-project/executor:v1.9.1` and then modifying `build_kaniko_image.sh` like the following.
```
	commandArgs := []string{
		// For some reason, kaniko is unable to detect it is running in a container when running on Fargate, so we need
		// to pass the force flag to ensure it runs.
		// Started Enabling caching here
		"--force",
		"--cache=true",
		"--cache-copy-layers=false",
		// https://github.com/GoogleContainerTools/kaniko/issues/2065
		"--cache-run-layers=true",
		"--compressed-caching=false",
	}
```

I then modified the `build_kaniko_image.sh` script fro the shared folder rin my infra/live repo to the following and ran it to build a new kaniko image
```
#!/bin/bash
# Build script for building Docker Container Images via the Gruntwork ECS deploy runner. Under the hood this
# script will invoke the docker-image-builder container of the ECS deploy runner (via the infrastructure-deployer CLI)
# that will invoke Kaniko to build the images.
#
# This script is intended to be run on a CI server that has the necessary credentials to invoke the ECS Deploy Runner in
# the Shared Services AWS account.
#
# This is the build script for the Kaniko Docker Image Builder Docker image. You can view the Dockerfile at the following URL:
# https://github.com/gruntwork-io/terraform-aws-ci/blob/v0.50.6/modules/ecs-deploy-runner/docker/kaniko

set -e


# readonly DOCKERFILE_REPO="https://github.com/gruntwork-io/terraform-aws-ci.git"  #Set back to this once it natively supports caching
readonly DOCKERFILE_REPO="https://github.com/YOUR_FORK_HERE/terraform-aws-ci.git"
readonly DOCKERFILE_REPO_REF="v0.50.4-catch"
readonly DOCKERFILE_CONTEXT_PATH="modules/ecs-deploy-runner/docker/kaniko"
readonly DEPLOY_RUNNER_REGION="us-east-1"
readonly ECR_REPO_REGION="us-east-1"
readonly ECR_REPO_NAME="kaniko"

# The account IDs where the AMI should be shared.
git_repo_root="$(git rev-parse --show-toplevel)"

function run {
  # Validate that the AMI is being built in the Shared Services account.
  local account_id
  account_id="$(aws sts get-caller-identity --query 'Account' --output text)"
  shared_account_id="$(jq -r '."shared".id' "$git_repo_root/accounts.json")"
  if [[ "$account_id" != "$shared_account_id" ]]; then
    >&2 echo "Not authenticated to the correct account. Expected: $shared_account_id ; Actual: $account_id"
    exit 1
  fi

  local -r ecr_repo_url="$shared_account_id"'.dkr.ecr.'"$ECR_REPO_REGION"'.amazonaws.com/'"$ECR_REPO_NAME"

    DOCKER_BUILDKIT=1 docker build --platform linux/amd64 --secret id=github-token,env=GITHUB_OAUTH_TOKEN -t "$ecr_repo_url:$DOCKERFILE_REPO_REF" .
}

# Run the main function if this script is called directly, instead of being sourced.
if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
  run "$@"
fi
```

I also created an additional ECR repo with a` /cache` suffix for kaniko to push the layers to.

I couldn't rely on the infra deployer script to build this remotely because of this change https://github.com/gruntwork-io/terraform-aws-ci/pull/495/files which I could not figure out.  

I would appreciate any guidance on how to update that correctly given https://github.com/gruntwork-io/terraform-aws-ci/pull/495/files so I don't have to rely on a local build in the future.

**Comment by zackleman** [comment]
This change reduced my build time from  ~10 minutes to ~3 minutes. Locally on my M1 machine docker can build in 5 seconds and 60 seconds when option  `--platform linux/amd64` is used.
